### PR TITLE
[refactor] support compile model and loss separately

### PR DESCRIPTION
**Pull request by tianyu-l**
Creating a new field in `JobConfig`, with the default being
```
[compile]
enable=false
components = ["model", "loss"]
```

This way we get to compile loss separately to get memory reduction, even when the model is not ready to be compiled.

This PR also applies loss compilation to DeepSeek 16B and 671B.

### CODE REVIEW

File: torchtitan/experiments/llama4/infra/parallelize.py [hunk 1]
```diff
@@ -62,10 +62,13 @@ def parallelize_llama(
...
+    model_compile_enabled = (
+        job_config.compile.enable and "model" in job_config.compile.components
+    )
     if parallel_dims.tp_enabled:
         if (
             job_config.parallelism.enable_async_tensor_parallel
-            and not job_config.training.compile
+            and not model_compile_enabled
         ):
             raise RuntimeError("Async TP requires --training.compile")
```
Comment (wwwjn) [inline torchtitan/experiments/llama4/infra/parallelize.py]: Need to fix the error message

File: torchtitan/experiments/simple_fsdp/parallelize.py [hunk 2]
```diff
@@ -37,10 +37,13 @@ def parallelize_llama(
...
+    model_compile_enabled = (
+        job_config.compile.enable and "model" in job_config.compile.components
+    )
     if parallel_dims.tp_enabled:
         if (
             job_config.parallelism.enable_async_tensor_parallel
-            and not job_config.training.compile
+            and not model_compile_enabled
         ):
             raise RuntimeError("Async TP requires --training.compile")
```
Comment (wwwjn) [inline torchtitan/experiments/simple_fsdp/parallelize.py]: Need to fix the error message as well

File: torchtitan/models/deepseek_v3/train_configs/deepseek_v3_16b.toml [hunk 3]
```diff
@@ -50,7 +49,7 @@ tensor_parallel_degree = 1
 enable_async_tensor_parallel = false
 pipeline_parallel_degree = 1
 pipeline_parallel_schedule = "Interleaved1F1B"
-expert_parallel_degree = 1
+expert_parallel_degree = 8
```
Comment (wwwjn) [inline torchtitan/models/deepseek_v3/train_configs/deepseek_v3_16b.toml:52]: If we encourage user use EP=8 for deepseek_16b model, we should also encourage them use EP=8 in 671B model

File: torchtitan/models/llama3/infra/parallelize.py [hunk 4]
```diff
@@ -63,10 +63,13 @@ def parallelize_llama(
...
+    model_compile_enabled = (
+        job_config.compile.enable and "model" in job_config.compile.components
+    )
     if parallel_dims.tp_enabled:
         if (
             job_config.parallelism.enable_async_tensor_parallel
-            and not job_config.training.compile
+            and not model_compile_enabled
         ):
             raise RuntimeError("Async TP requires --training.compile")
```
Comment (wwwjn) [inline torchtitan/models/llama3/infra/parallelize.py]: Same here

File: torchtitan/models/llama3/train_configs/llama3_405b.toml [hunk 5]
```diff
@@ -56,6 +55,10 @@ async_mode = "disabled" # ["disabled", "async", "async_with_pinned_mem"]
 [activation_checkpoint]
 mode = "full" # ["none", "selective", "full"]
```
Comment (wwwjn) [inline torchtitan/models/llama3/train_configs/llama3_405b.toml:56]: Should we change to selective AC for 405b model and 70b model as well?

File: torchtitan/models/deepseek_v3/train_configs/deepseek_v3_16b.toml (hunk 3 above)
Comment (tianyu-l) [reply torchtitan/models/deepseek_v3/train_configs/deepseek_v3_16b.toml:52]: No because 16B can run with EP8, but for 671B EP8 is not enough, we probably need EP32. It has to be benchmarked before we have update this field.

File: torchtitan/models/llama3/train_configs/llama3_405b.toml (hunk 5 above)
Comment (tianyu-l) [reply torchtitan/models/llama3/train_configs/llama3_405b.toml:56]: No because the `selective` policy is to avoid recomputing the d2h sync and a2a which only exists in MoE models, being added in https://github.com/pytorch/torchtitan/pull/1597
For dense models, `full` seems good.

**Review by wwwjn — APPROVED** [review]
LGTM!
